core/src/main/java/com/linecorp/armeria/internal/Http2KeepAliveHandler.java

```diff
@@ -0,0 +1,276 @@
+
+        if (shutdownFuture != null) {
+            final boolean isCancelled = shutdownFuture.cancel(false);
+            if (!isCancelled) {
+                logger.debug("{} shutdownFuture cannot be cancelled because of late PING ACK", channel);
+            }
+        }
+        logger.debug("{} Received PING(ACK=1) in {} ns", channel, elapsed);
+        state = State.IDLE;
+    }
+
+    private boolean isGoodPingAck(long data) {
+        // This condition can be true when channel read some data other than PING ACK frame
+        // or a PING ACK is received without sending PING in first place.
+        if (state != State.PENDING_PING_ACK) {
+            logger.debug("{} Ignoring PING(ACK=1, DATA={}) received", channel, data);
+            return false;
+        }
+        if (lastPingPayload != data) {
+            logger.debug("{} Unexpected PING(ACK=1, DATA={}) received, " +
+                         "but expecting PING(ACK=1, DATA={})", channel, data, lastPingPayload);
+            return false;
+        }
+        return true;
+    }
+
+    @VisibleForTesting
+    State state() {
+        return state;
+    }
+
+    @VisibleForTesting
+    long lastPingPayload() {
+        return lastPingPayload;
+    }
+
+    private void closeChannelAndLog() {
+        if (state == State.SHUTDOWN) {
+            return;
+        }
+        logger.debug("{} Closing an idle channel", channel);
+        channel.close().addListener(future -> {
+            if (future.isSuccess()) {
+                logger.debug("{} Closed an channel", channel);
+            } else {
+                logger.debug("{} Failed to close an idle channel", channel, future.cause());
+            }
+            state = State.SHUTDOWN;
+        });
+    }
+
+    private long getStopwatchElapsedInNanos() {
+        if (stopwatch == null) {
+            return -1;
+        }
+        return stopwatch.elapsed(TimeUnit.NANOSECONDS);
+    }
+
+    /**
+     * State changes from IDLE -> PING_SCHEDULED -> PENDING_PING_ACK -> IDLE and so on. When the
+     * channel is inactive then the state changes to SHUTDOWN.
+     */
+    enum State {
+        /* Nothing happening, but waiting for IdleStateEvent */
+        IDLE,
+
+        /* PING is scheduled */
+        PING_SCHEDULED,
+
+        /* PING is sent and is pending ACK */
+        PENDING_PING_ACK,
+
+        /* Not active anymore */
+        SHUTDOWN
+    }
+
+    private class PingWriteListener implements ChannelFutureListener {
+
+        @Override
+        public void operationComplete(ChannelFuture future) throws Exception {
+            if (future.isSuccess()) {
+                final EventLoop el = channel.eventLoop();
+                shutdownFuture = el.schedule(shutdownRunnable, pingTimeoutMillis, TimeUnit.MILLISECONDS);
+                state = State.PENDING_PING_ACK;
+                resetStopwatch();
+            } else {
+                // Mostly because the channel is already closed. So ignore and change state to IDLE.
```

Comment:
Perhaps we should not log anything if the write future has been cancelled by cancelFutures()?